src/epm_dynamic.rs

```diff
@@ -0,0 +1,213 @@
+		MAX_VOTES_PER_VOTER.to_string(),
+		max_votes_per_voter
+	);
+
+	static_types::MaxWeight::set(max_weight);
+	static_types::MaxLength::set(max_length);
+	static_types::MaxVotesPerVoter::set(max_votes_per_voter);
+
+	Ok(())
+}
+
+fn invalid_metadata_error<E: std::error::Error>(item: String, err: E) -> Error {
+	Error::InvalidMetadata(format!("{} failed: {}", item, err))
+}
+
+fn read_constant<'a, T: serde::Deserialize<'a>>(
+	api: &SubxtClient,
+	constant: EpmConstant,
+) -> Result<T, Error> {
+	let (epm_name, constant) = constant.to_parts();
+
+	let val = api
+		.constants()
+		.at(&subxt::dynamic::constant(epm_name, constant))
+		.map_err(|e| invalid_metadata_error(constant.to_string(), e))?;
+
+	scale_value::serde::from_value::<_, T>(val).map_err(|e| {
+		Error::InvalidMetadata(format!("Decoding `{}` failed {}", std::any::type_name::<T>(), e))
+	})
+}
+
+pub fn signed_solution<S: NposSolution + Encode + TypeInfo + 'static>(
+	solution: RawSolution<S>,
+) -> Result<DynamicTxPayload<'static>, Error> {
+	let scale_solution = to_scale_value(solution).map_err(|e| {
+		Error::DynamicTransaction(format!("Failed to decode `RawSolution`: {:?}", e))
+	})?;
+
+	Ok(subxt::dynamic::tx(EPM_PALLET_NAME, "submit", vec![scale_solution]))
+}
+
+pub fn unsigned_solution<S: NposSolution + Encode + TypeInfo + 'static>(
+	solution: RawSolution<S>,
+	witness: SolutionOrSnapshotSize,
+) -> Result<DynamicTxPayload<'static>, Error> {
+	let scale_solution = to_scale_value(solution)?;
+	let scale_witness = to_scale_value(witness)?;
+
+	Ok(subxt::dynamic::tx(EPM_PALLET_NAME, "submit_unsigned", vec![scale_solution, scale_witness]))
+}
+
+pub async fn signed_submission_at<S: NposSolution + Decode + TypeInfo + 'static>(
+	idx: u32,
+	at: Hash,
+	api: &SubxtClient,
+) -> Result<Option<SignedSubmission<S>>, Error> {
+	let scale_idx = Value::u128(idx as u128);
+	let addr = subxt::dynamic::storage(EPM_PALLET_NAME, "SignedSubmissionsMap", vec![scale_idx]);
+
+	match api.storage().fetch(&addr, Some(at)).await {
+		Ok(Some(val)) => {
+			let val = val.remove_context();
+			let v = encode_scale_value::<SignedSubmission<S>>(&val)?;
+			let submissions = Decode::decode(&mut v.as_ref())?;
+			Ok(Some(submissions))
+		},
+		Ok(None) => Ok(None),
+		Err(err) => Err(err.into()),
+	}
+}
+
+fn make_type<T: scale_info::TypeInfo + 'static>() -> (TypeId, PortableRegistry) {
+	let m = scale_info::MetaType::new::<T>();
+	let mut types = scale_info::Registry::new();
+	let id = types.register_type(&m);
+	let portable_registry: PortableRegistry = types.into();
+
+	(id.into(), portable_registry)
+}
+
+fn to_scale_value<T: scale_info::TypeInfo + 'static + Encode>(val: T) -> Result<Value, Error> {
+	let (ty_id, types) = make_type::<T>();
+
+	let bytes = val.encode();
+
+	decode_as_type(&mut bytes.as_ref(), ty_id, &types)
+		.map(|v| v.remove_context())
+		.map_err(|e| {
+			Error::DynamicTransaction(format!(
+				"Failed to decode {}: {:?}",
+				std::any::type_name::<T>(),
+				e
+			))
+		})
+}
```

Comment:
Ooh, I see, this is a clever way to take some type implementing `TypeInfo + Encode` into a `scale_value::Value`! It's a shame to have to encode it to bytes and then decode from the bytes, but offhand I can't think of a better/more reliable way.

You _can_ go from a `T: Serialize` straight into a `scale_value::Value`, which would be more efficient (no intermediate step), but it also goes through the serde data model. Might be worth trying if you wanted faster and fewer lines of code, but the downside is that the serde and scale impls don't necessarily need to agree with eachother (eg H256 serializes to a string, which would turn into a scale_value primitive String rather than a sequence of bytes, and the string then wouldn't encode to bytes as would be needed to encode it properly).

Reply:
yeah but these types doesn't implement `serde` maybe I can convince @kianenigma to add  `deserialize/serialize` on these types :)

Reply:
Makes sense! Aside from using Serde, or some other similar machinery, I think what you've done is the best possible approach anyways :)